File: 1.HTTP-AND-CORS/code/typescript/large_downloads_server_client_streaming_14.ts
import express, { Request, Response } from 'express';
import multer from 'multer';

const app = express();

// 1. Receive a multipart upload
// up to 32 MB in memory, overflow spills to disk (handled by multer limits/disk storage)
const upload = multer({ dest: 'uploads/', limits: { fileSize: 32 * 1024 * 1024 } });

app.post('/upload', upload.single('file'), (req: Request, res: Response) => {
    const file = req.file;
    if (!file) {
        res.status(400).send('no file');
        return;
    }
    res.send(`received ${file.originalname}`);
});

// 2. Stream a response in chunks (Server-Sent Events)
app.get('/stream', (req: Request, res: Response) => {
    res.setHeader('Content-Type', 'text/event-stream');
    res.setHeader('Connection', 'keep-alive');
    
    let i = 0;
    const interval = setInterval(() => {
        res.write(`data: chunk ${i}\n\n`); // write() pushes each chunk to the client immediately
        i++;
        if (i >= 5) {
            clearInterval(interval);
            res.end();
        }
    }, 1000);
});

File: 1.HTTP-AND-CORS/code/rust/large_downloads_server_client_streaming_14.rs
use actix_web::{web, HttpResponse, Responder};
use futures_util::stream::StreamExt as _;
use std::time::Duration;

// 1. Receive a multipart upload (stubbed for conceptual translation)
async fn upload(mut payload: web::Payload) -> impl Responder {
    // Actix payload handles streaming data for multipart
    let mut bytes = web::BytesMut::new();
    while let Some(item) = payload.next().await {
        bytes.extend_from_slice(&item.unwrap());
    }
    HttpResponse::Ok().body("received file")
}

// 2. Stream a response in chunks (Server-Sent Events)
async fn stream() -> impl Responder {
    let stream = async_stream::stream! {
        for i in 0..5 {
            yield Ok::<_, actix_web::Error>(web::Bytes::from(format!("data: chunk {}\n\n", i)));
            tokio::time::sleep(Duration::from_secs(1)).await;
        }
    };
    
    HttpResponse::Ok()
        .content_type("text/event-stream")
        .insert_header(("Connection", "keep-alive"))
        .streaming(stream)
}
